File: app/src/main/java/com/castelijns/mmdemo/photos/PhotosPresenter.java
package com.castelijns.mmdemo.photos;

import android.util.SparseArray;
import android.widget.ImageView;

import com.castelijns.mmdemo.models.Photo;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class PhotosPresenter implements PhotosContract.Presenter {

    private PhotosContract.View view;
    private PhotosRepo repo;

    PhotosPresenter(PhotosContract.View view) {
        this.view = view;
        repo = PhotosRepo.getInstance();
    }

    @Override
    public void start() {
        view.clearList();
        view.showLoading();

        repo.getAllPhotos()
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.computation())
                .map(photos -> {
                    // Split albums by albumId.
                    SparseArray<List<Photo>> albumPhotos = new SparseArray<>();
                    for (Photo photo : photos) {
                        List<Photo> photoList = albumPhotos.get(photo.getAlbumId(), new ArrayList<>());
                        photoList.add(photo);

                        albumPhotos.append(photo.getAlbumId(), photoList);
                    }
                    return albumPhotos;
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<SparseArray<List<Photo>>>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(SparseArray<List<Photo>> albumPhotos) {
                        view.showPhotos(albumPhotos);

                        int photoCount = 0;
                        for (int i = 0; i < albumPhotos.size(); i++) {
                            photoCount += albumPhotos.valueAt(i).size();
                        }
                        view.showPhotoCount(photoCount, albumPhotos.size());
                    }

                    @Override
                    public void onError(Throwable e) {
                        view.hideLoading();
                        view.showError();
                    }

                    @Override
                    public void onComplete() {
                        view.hideLoading();
                    }
                });
    }

    @Override
    public void start(int albumId) {
        filterAlbums(albumId);
    }

    @Override
    public void stop() {

    }

    @Override
    public void onPhotoClicked(Photo photo, ImageView ivPhoto) {
        view.showPhotoDetail(photo, ivPhoto);
    }

    @Override
    public void filterAlbums(int albumId) {
        view.clearList();
        view.showLoading();

        repo.getAllPhotosForAlbumId(albumId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<List<Photo>>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(List<Photo> photos) {
                        if (photos.size() > 0) {
                            SparseArray<List<Photo>> albumPhotos = new SparseArray<>();
                            albumPhotos.append(photos.get(0).getAlbumId(), photos);

                            view.showPhotos(albumPhotos);
                            view.showPhotoCount(photos.size(), 1);
                        } else {
                            view.showPhotos(new SparseArray<>());
                            view.showPhotoCount(0, 0);
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        view.hideLoading();
                        view.showError();
                    }

                    @Override
                    public void onComplete() {
                        view.hideLoading();
                    }
                });
    }
}

File: app/src/main/kotlin/com/castelijns/mmdemo/photos/PhotosPresenter.kt
package com.castelijns.mmdemo.photos

import android.widget.ImageView
import com.castelijns.mmdemo.albums.AlbumsRepo
import com.castelijns.mmdemo.app.BasePresenter
import com.castelijns.mmdemo.models.Album
import com.castelijns.mmdemo.models.Photo
import kotlinx.coroutines.experimental.launch

class PhotosPresenter(
        private val view: PhotosContract.View
) : BasePresenter() {

    private val photosRepo = PhotosRepo
    private val albumsRepo = AlbumsRepo

    override fun start() {
        view.clearList()
        view.showLoading()

        launch {
            try {
                loadPhotos()
            } catch (e: Exception) {
                view.showError()
            } finally {
                view.hideLoading()
            }
        }
    }

    private suspend fun loadPhotos() {
        val allPhotosJob = photosRepo.getAllPhotos()
        val allAlbumsJob = albumsRepo.getAllAlbums()

        val photos = allPhotosJob.await()
        val albums = allAlbumsJob.await()

        // Assign album titles to photos
        photos.forEach { photo ->
            photo.albumTitle = albums.first { album -> album.id == photo.albumId }.title
        }

        // Split photos by albumId
        val photosByAlbum: Map<Album, List<Photo>> = photos.groupBy { photo ->
            albums.first { album -> album.id == photo.albumId }
        }

        view.showPhotosByAlbum(photosByAlbum)
        view.showPhotoCount(photos.size, photosByAlbum.keys.size)

        photosRepo.cacheData(photos)
        albumsRepo.cacheData(albums)
    }

    fun start(albumId: Int) {
        view.clearList()
        view.showLoading()

        launch {
            try {
                loadPhotosForAlbum(albumId)
            } catch (e: Exception) {
                view.showError()
            } finally {
                view.hideLoading()
            }
        }
    }

    fun onPhotoClicked(photo: Photo, ivPhoto: ImageView) {
        view.showPhotoDetail(photo, ivPhoto)
    }

    private suspend fun loadPhotosForAlbum(albumId: Int) {
        val photosJob = photosRepo.getAllPhotosForAlbumId(albumId)
        val albumJob = albumsRepo.getAlbumById(albumId)

        val photos = photosJob.await()
        val album = albumJob.await()

        // Assign album title to photos
        photos.forEach { photo ->
            photo.albumTitle = album.title
        }

        if (photos.isNotEmpty()) {
            view.showPhotosByAlbum(mapOf(album to photos))
            view.showPhotoCount(photos.size, 1)
        } else {
            view.showPhotosByAlbum(mapOf())
            view.showPhotoCount(0, 0)
        }
    }

}
